app/src/main/java/org/fossasia/susi/ai/skills/groupwiseskills/GroupWiseSkillsFragment.kt

```diff
@@ -93,6 +96,15 @@ class GroupWiseSkillsFragment : Fragment(), IGroupWiseSkillsView, SwipeRefreshLa
         skillsAdapter.notifyDataSetChanged()
     }
 
+    override fun onAttach(context: Context) {
+        super.onAttach(context);
+        try {
+            skillCallback = activity as SkillFragmentCallback
+        } catch (e: ClassCastException) {
+            throw ClassCastException(activity.toString() + " must implement SkillFragmentCallback")
```

Comment:
The activity with which the GroupWiseSkillsFragment is attached is SkillsActivity and SkillFragmentCallback interface is used to load skills details when a skill card is clicked. This is what is leading to this issue. I am not able to understand how to handle two types of contexts?

Reply:
What if a separate activity is used in future to attach this fragment, or an espresso/roboelectric test is written to test this Fragment which cannot extend this class, so rather than being able to test this fragment, it will throw ClassCastException

Reply:
You just have to check if and only if the activity is of type SklllFragmentCallback, only then assign it in the callback variable

This is a standard process of assigning cast classes in Java

Reply:
@iamareebjamal But, activity will never be of type SkillFragmentCallback by itself. Thus, when I use 'if' to check if the activity is of type SkillFragmentCallback, the skills are never fetched. The SkillFragmentCallback interface is used to load group wise skills as well as skill details. So, 'skillCallback' will always be used if this fragment is attached. But, this is also true that ClassCastException might occur. Also, I checked some answers on StackOverflow to see how we can handle this kind of a case, but I see everywhere they have used explicit type casting to assign this variable. Also, I read this official documentation [Communicating with other Fragments](https://developer.android.com/training/basics/fragments/communicating), here also the activity has been explicitly cast to the interface.
I am also concerned about the crashes that might occur in future because of this, so is there something that I am missing on or maybe any other way we can implement this?

Reply:
I'm sure you know about the hazards of copy pasting code from docs and stackoverflow, they use minimal code to get the idea accross, if they follow every best practices and design pattern used in production code, every three liner would become Factory, Builder, Singleton

Reply:
@iamareebjamal  I didn't copy paste the code from there. But, I checked these forums to look where I was stuck or maybe if somebody has faced a similar issue. Thus, I quoted it. I will try to check it again.

Reply:
@iamareebjamal GroupWiseSkillsFragment is not implementing the SkillFragmentCallback interface. This interface has already been implemented in the SkillsActivity. A lateinit var of type SkillFragment has been defined, as also done in SkillsListingFragment, which is passed in the adapter and used to load only one of the two functions i.e. to load loadSkillDetails. Hence, it is not possible to check if the context is of type SkillFragmentCallback here. I am unable to understand how to resolve your review request by checking the context type. As I have used an existing activity, I have followed a similar implementation as already done for SkillsListingFragment. It would be very helpful if you could please help me a bit with this as this is blocking further development.

Reply:
Save this code as `inheritance.kts` and run with `kotlinc -script inheritance.kts` 

Follow the instructions written in it, play around with it for some time

```kotlin
interface Greeter {
    fun greet()
}

open class Person(val name: String)

class PersonGreeter(name: String): Person(name), Greeter {

    override fun greet() {
        println("Hello ${this.name}")
    }

}

val personGreeter: Person = PersonGreeter("Areeb")
val person: Person = Person("Areeb")


(personGreeter as Greeter).greet() // unsafe, you are just getting person in Person variable
// There is no surity that Person variable is of type Greeter as well, -> 
// comment the following line one you are convinced that this is unsafe
(person as Greeter).greet() // This will throw exception

// This is basic Java, you must know it by now!
// To safely resolve a variable to a case, you must check its instance

if (personGreeter is Greeter) {
    // Wow! Kotlin has automatically cast personGreeter as Greeter
    personGreeter.greet()
} else {
    println("personGreeter is not a Greeter")
}

if (person is Greeter) {
    // This will probably not get executed and not throw ClassCastException
    person.greet()
} else {
    println("person is not a Greeter")
}
```

Refer these for further reading
https://kotlinlang.org/docs/reference/classes.html
https://kotlinlang.org/docs/reference/interfaces.html

Reply:
If same approach is being used elsewhere in the project, please raise appropriate issues and fix those using similar approach once you are convinced that this is the proper way to handle casting

Reply:
Sure @iamareebjamal , I will keep that in mind. Thanks. :)
Checking the above links.

Reply:
Thanks @iamareebjamal this was very helpful. I tried running the above code and commented out a block at a time and also tried modifying it, I realised that the earlier implementation was not correct and there were high chances of getting a ClassCastException. Please have a look at the modified code.